README.md
# Rack::JQueryUI::Themes #

[jQuery-UI](http://jqueryui.com/) ***themes*** CDN script tags and fallback in one neat package.

### Build status ###

Master branch:
[![Build Status](https://travis-ci.org/yb66/rack-jquery_ui-themes.png?branch=master)](https://travis-ci.org/yb66/rack-jquery_ui-themes)

### Why? ###

I get tired of copy and pasting and downloading and moving… jQuery files and script tags etc. This does it for me (along with https://github.com/yb66/rack-jquery), and keeps version management nice 'n' easy.

### See also ###

[Rack::JQuery](https://github.com/yb66/rack-jquery)  
[Rack::JQueryUI](https://github.com/yb66/rack-jquery_ui)

They're not dependencies, but if you're going to use this then I bet you'll be interested in using them.

### What's with the plural ###

Rack::JQueryUI::Theme doesn't sound right to me, as it's accessing the _Themes_ service/files. I doubt you'll ever instantiate a instance of the class yourself, and the computer doesn't mind, so let it go.

### Usage ###

Have a look in the examples directory, but here's a snippet.

* Install it (see below)
* `require 'rack/jquery_ui/themes'`.
* Put this in your middleware stack: `use Rack::JQueryUI::Themes, :theme => "vader"`
* Put this in the head of your layout (the example is Haml but you can use whatever you like)

    <pre><code>
    %head
      = Rack::JQueryUI::Themes.cdn env, :organisation => :microsoft
    </code></pre>

Now you have the script tags to Microsoft's CDN in the head (you can also use Media Temple or Microsoft, see the docs).

It also adds in a bit of javascript that will load in a locally kept version of jQuery, just incase the CDN is unreachable. The script will use the "/js/jquery-ui/1.10.3/themes/:THEME/jquery-ui.min.css" path (or, instead of 1.10.3, whatever is in {Rack::JQueryUI::JQUERY_UI_VERSION}), where `:THEME` is the name of the theme you specified, the default being `smoothness`.

That was easy.

### Note ###

You have to have loaded jQuery _before_ using the CDN helper, as Rack::jQueryUI::Themes relies on it. I've already mentioned [Rack::JQuery](https://github.com/yb66/rack-jquery) which you can use to do this, or load the script however you like. Just remember that it needs to be there.

### Version numbers ###

This library uses [semver](http://semver.org/) to version the **library**. That means the library version is ***not*** an indicator of quality but a way to manage changes. The version of jQuery-UI can be found in the lib/rack/jquery_ui/themes/version.rb file, or via the {Rack::JQueryUI::JQUERY_UI_VERSION} constant.

### Installation ###

Add this line to your application's Gemfile:

    gem 'rack-jquery_ui-themes'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rack-jquery_ui-themes

### Contributing ###

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

### Licence ###

See the LICENCE.txt file.

CHANGES.md
# CH CH CH CHANGES #

## Tuesday the 18th of August 2015, v3.0.1 ##

* Correct typo in docs.
* Better git ignores.

----


## Tuesday the 6th of August 2013, v3.0.0 ##

* Removed class methods for theme and themes, they just don't work with the Rack architecture, at least not the way I'm using them.
* Found a way to check if the external CDN has loaded, by adding a meta tag and then one of the jQuery UI classes to it, and seeing if it changes.
* Added some extra keys the internal paths hash to help ascertain which theme is being called.
* Added a fallback option to the cdn method, mainly to help with testing but also because it might be wanted.

----


## Friday the 19th of July 2013, v2.1.1 ##

* Removed rogue console logging statement in the fallback javascript.

----


## Thursday the 18th of July 2013, v2.1.0 ##

* Bumped version of jQuery UI back up to 1.10.3.

----


## Thursday the 18th of July 2013, v2.0.0 ##

* The fallback wasn't working properly, fixed.
* Got one of the jQuery team to clarify which parts of the themes download was needed, so updated that.
* Added new option for selecting several themes.
* Fixed some bad architectural choices with the code.
* Downgraded to version 1.10.1 of jQuery UI to get this fixed, and the next release will bump it back up to 1.10.3.
* Made Media Temple the default CDN, because Google and Microsoft have sided with Satan over all this PRISM/TEMPURA nonsense, and I don't feel like making it easier for them to track people around the web.
* This is all breaking!!! so the major version has been bumped.

----


## v1.0.0 ##

* Bumped to 1.1.0 for semver.
* Updated to version 1.10.3 of jQuery UI.
* 100% documentation coverage.
* 100% code coverage.

----


## v0.1.1 ##

Wednesday the 24th of April 2013

* Fixed clobbering of extra call method.

____


## v0.1.0 ##

Friday the 22nd of March 2013

* Made call thread safe by duplicating `call` method.

----

## v0.0.3 ##

Friday the 8th of March 2013

* Updated docs again.

----

## v0.0.2 ##

Friday the 8th of March 2013

* Updated docs.

----

## v0.0.1 ##

Friday the 8th of March 2013

* First release. Handles CDN for jQuery-UI themes, and provides a fallback to local CSS and image files.

----


spec/spec_helper.rb
# encoding: UTF-8

require 'rspec'
Spec_dir = File.expand_path( File.dirname __FILE__ )


# code coverage
require 'simplecov'
SimpleCov.start do
  add_filter "/vendor/"
  add_filter "/bin/"
  add_filter "/spec/"
  add_filter "/examples/"
end

require "rack/test"
ENV['RACK_ENV'] ||= 'test'
ENV["EXPECT_WITH"] ||= "racktest"


Dir[ File.join( Spec_dir, "/support/**/*.rb")].each do |f|
  require f
end
                                  

module RackJQueryUIThemesHelpers
  def subber( s, opts={} )
    script_name = opts.fetch("SCRIPT_NAME","/")
    fallback_url = ::File.join script_name, "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/#{Rack::JQueryUI::Themes::JQUERY_UI_THEME_FILE}"
    theme = opts.fetch :theme, "smoothness"
    cdn = opts[:cdn]
    s = s.gsub(/\:CDNURL/, cdn) if cdn
    s.sub! /\:FALLBACK_URL/, fallback_url
    s.gsub(/\:THEME/, theme)
  end
end

RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.include RackJQueryUIThemesHelpers
end


spec/themes_spec.rb
# encoding: UTF-8

require 'spec_helper'
require_relative "../lib/rack/jquery_ui/themes.rb"

shared_examples "Given an organisation and a theme" do
  let(:link) { "<link rel='stylesheet' href='#{cdn}' type='text/css' />" }
  let(:expected) { "#{subber link, opts}\n#{subber Rack::JQueryUI::Themes::FALLBACK, opts}" }
  it { should == expected }
end

shared_context "Calling the CDN" do
  subject { Rack::JQueryUI::Themes.cdn env, opts }
end


describe "The class methods" do
  context "Given an organisation" do
    subject { Rack::JQueryUI::Themes.cdn env, opts }
    let(:theme) { "smoothness" }
    let(:opts) {
      {:theme => theme, :cdn => cdn, :organisation => organisation}
    }
    let(:env) {
      {}
    }
    context "of nil (the default)" do
      let(:organisation) { nil }
      let(:cdn) { Rack::JQueryUI::Themes::CDN::MEDIA_TEMPLE }
      include_examples "Given an organisation and a theme"
      context "and given a theme" do
        context "That is valid" do
          let(:theme) { Rack::JQueryUI::Themes::STANDARD_THEMES.sample}
          include_context "Calling the CDN"
          include_examples "Given an organisation and a theme"
        end
        context "That is not valid" do
          context "Because it's not in the standard themes" do
            let(:theme) { "breakfastsupercharger" }
            it "should fail" do
              expect{ Rack::JQueryUI::Themes.cdn organisation, opts }.to raise_error
            end
          end
          context "Because different themes were chosen at set up" do
            let(:app) {
              Sinatra.new do
                use Rack::JQuery
                use Rack::JQueryUI
                use Rack::JQueryUI::Themes, :themes => %w{pepper-grinder swanky-purse trontastic}
                get('/') { Rack::JQueryUI::Themes.cdn organisation, opts }
              end
            }
            let(:theme) { "overcast" }
            it "should fail" do
              expect{ get "/" }.to raise_error
            end
          end
        end
      end
    end
    context "of :google" do
      let(:organisation) { :google }
      let(:cdn) { Rack::JQueryUI::Themes::CDN::GOOGLE }
      include_examples "Given an organisation and a theme"
    end
    context "of :microsoft" do
      let(:organisation) { :microsoft }
      let(:cdn) { Rack::JQueryUI::Themes::CDN::MICROSOFT }
      include_examples "Given an organisation and a theme"
    end
    context "of :media_temple" do
      let(:organisation) { :media_temple }
      let(:cdn) { Rack::JQueryUI::Themes::CDN::MEDIA_TEMPLE }
      include_examples "Given an organisation and a theme"
    end
  end
end

# Bit of internal checking, for sanity.
describe "Path_to_vendor" do
  subject { Rack::JQueryUI::Themes::Path_to_vendor }
  it { should == "../../../vendor/assets/javascripts/jquery-ui" }
end

describe "Inserting the CDN" do
  include_context "All routes"
  let(:opts) {
    {:theme => theme, :url => cdn}.delete_if{|k,v| v.nil?}
  }
  let(:expected) { subber cdn, opts  }
  context "Check the examples run at all" do
    before do
      get "/"
    end
    it_should_behave_like "Any route"
  end
  context "Given a theme" do
    context "That is valid" do
      let(:theme) { "vader" }
      context "Google CDN" do
        let(:cdn) { Rack::JQueryUI::Themes::CDN::GOOGLE }
        before do
          get "/google-cdn"
        end
        it_should_behave_like "Any route"
        subject { last_response.body }
        it { should include expected }
      end
      context "Microsoft CDN" do
        before do
          get "/microsoft-cdn"
        end
        it_should_behave_like "Any route"
        subject { last_response.body }
        let(:cdn) { Rack::JQueryUI::Themes::CDN::MICROSOFT }
        it { should include expected }
      end
      context "Media_temple CDN" do
        before do
          get "/media-temple-cdn"
        end
        it_should_behave_like "Any route"
        subject { last_response.body }
        let(:cdn) { Rack::JQueryUI::Themes::CDN::MEDIA_TEMPLE }
        it { should include expected }
      end
      context "Unspecified CDN" do
        before do
          get "/"
        end
        it_should_behave_like "Any route"
        subject { last_response.body }
        let(:cdn) { Rack::JQueryUI::Themes::CDN::MEDIA_TEMPLE }
        it { should include expected }
      end
    end
    context "that is not valid" do
      let(:app) {
        Sinatra.new do
          use Rack::JQuery
          use Rack::JQueryUI
          use Rack::JQueryUI::Themes, :theme => "breakfastsupercharger"
          get('/') { "Shouldn't reach this" }
        end
      }
      it "should fail" do
        expect{ get "/" }.to raise_error
      end
    end
  end
end


require 'timecop'
require 'time'

describe "Serving the fallback jquery" do
  include_context "All routes"
  let(:theme) { {:theme => "vader"} }
  let(:cdn) { Rack::JQueryUI::Themes::CDN::MEDIA_TEMPLE }
  let(:opts) { theme.merge( :cdn => cdn ) }
  context "Single request" do
    context "for route that doesn't exist" do
      let(:url) { "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/blah/#{Rack::JQueryUI::Themes::JQUERY_UI_THEME_FILE}"}
      before do
        get url
      end
      subject { last_response.status }
      it { should == 404 }
    end
    context "for CSS" do
      let(:url) { subber "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/#{Rack::JQueryUI::Themes::JQUERY_UI_THEME_FILE}", opts }
      before do
        get url
      end
      it_should_behave_like "Any route"
      subject { last_response.body }
      it { should start_with "/*! jQuery UI - v#{Rack::JQueryUI::JQUERY_UI_VERSION}" }
    end
    context "for an image" do
      ex_path = File.expand_path "../../vendor/assets/javascripts/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/vader/images", __FILE__
      img = Dir.new( ex_path ).entries.find{|f| f =~ /\.png$/ }
      let(:url) {  "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/vader/images/#{img}" }
      before do
        get url
      end
      it_should_behave_like "Any route"
      subject { last_response.body }
      it { should_not be_nil }
    end
  end
  context "Re requests" do
    before do
      at_start = Time.parse(Rack::JQueryUI::JQUERY_UI_VERSION_DATE) + 60 * 60 * 24 * 180
      Timecop.freeze at_start
      get url
      Timecop.travel Time.now + 86400 # add a day
      get url, {}, {"HTTP_IF_MODIFIED_SINCE" => Rack::Utils.rfc2109(at_start) }
    end
    subject { last_response }
    context "for CSS" do
      let(:url) { subber "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/#{Rack::JQueryUI::Themes::JQUERY_UI_THEME_FILE}", opts }
      its(:status) { should == 304 }
    end
    context "for an image" do
      ex_path = File.expand_path "../../vendor/assets/javascripts/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/vader/images", __FILE__
      img = Dir.new( ex_path ).entries.find{|f| f =~ /\.png$/ }
      let(:url) {  "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/themes/vader/images/#{img}" }
      its(:status) { should == 304 }
    end
  end
end

spec/rack_jquery_ui_spec.rb
# encoding: UTF-8

require 'spec_helper'
require_relative "../lib/rack/jquery_ui.rb"

describe "The class methods" do
  subject { Rack::JQueryUI.cdn :organisation => organisation }
  context "Given an argument" do
    context "of nil (the default)" do
      let(:organisation) { nil }
      it { should == "#{Rack::JQueryUI::GOOGLE}\n#{Rack::JQueryUI::FALLBACK}" }
    end
    context "of :google" do
      let(:organisation) { :google }
      it { should == "#{Rack::JQueryUI::GOOGLE}\n#{Rack::JQueryUI::FALLBACK}" }
    end
    context "of :microsoft" do
      let(:organisation) { :microsoft }
      it { should == "#{Rack::JQueryUI::MICROSOFT}\n#{Rack::JQueryUI::FALLBACK}" }
    end
    context "of :media_temple" do
      let(:organisation) { :media_temple }
      it { should == "#{Rack::JQueryUI::MEDIA_TEMPLE}\n#{Rack::JQueryUI::FALLBACK}" }
    end
  end
end

describe "Inserting the CDN" do
  include_context "All routes"
  context "Check the examples run at all" do
    before do
      get "/"
    end
    it_should_behave_like "Any route"
  end
  context "Google CDN" do
    before do
      get "/google-cdn"
    end
    it_should_behave_like "Any route"
    subject { last_response.body }
    let(:expected) { Rack::JQueryUI::GOOGLE }
    it { should include expected }
  end
  context "Microsoft CDN" do
    before do
      get "/microsoft-cdn"
    end
    it_should_behave_like "Any route"
    subject { last_response.body }
    let(:expected) { Rack::JQueryUI::MICROSOFT }
    it { should include expected }
  end
  context "Media_temple CDN" do
    before do
      get "/media-temple-cdn"
    end
    it_should_behave_like "Any route"
    subject { last_response.body }
    let(:expected) { Rack::JQueryUI::MEDIA_TEMPLE }
    it { should include expected }
  end
  context "Unspecified CDN" do
    before do
      get "/unspecified-cdn"
    end
    it_should_behave_like "Any route"
    subject { last_response.body }
    let(:expected) { Rack::JQueryUI::GOOGLE }
    it { should include expected }
  end
end


require 'timecop'
require 'time'

describe "Serving the fallback jquery" do
  include_context "All routes"
  before do
    get "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/#{Rack::JQueryUI::JQUERY_UI_FILE_NAME}"
  end
  it_should_behave_like "Any route"
  subject { last_response.body }
  it { should start_with "/*! jQuery UI - v#{Rack::JQueryUI::JQUERY_UI_VERSION}" }

  context "Re requests" do
    before do
      at_start = Time.parse(Rack::JQueryUI::JQUERY_UI_VERSION_DATE) + 60 * 60 * 24 * 180
      Timecop.freeze at_start
      get "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/#{Rack::JQueryUI::JQUERY_UI_FILE_NAME}"
      Timecop.travel Time.now + 86400 # add a day
      get "/js/jquery-ui/#{Rack::JQueryUI::JQUERY_UI_VERSION}/#{Rack::JQueryUI::JQUERY_UI_FILE_NAME}", {}, {"HTTP_IF_MODIFIED_SINCE" => Rack::Utils.rfc2109(at_start) }
    end
    subject { last_response }
    its(:status) { should == 304 }
    
  end
end

examples/config.rb
require 'sinatra/base'
require 'haml'
require 'rack/jquery'
require 'rack/jquery_ui'
require 'rack/jquery_ui/themes'

class App < Sinatra::Base

  enable :inline_templates
  use Rack::JQuery
  use Rack::JQueryUI
  use Rack::JQueryUI::Themes, :theme => "vader", :themes => Rack::JQueryUI::Themes::STANDARD_THEMES

  get "/" do
    haml :index, :layout => :unspecified
  end

  get "/google-cdn" do
    haml :index, :layout => :google
  end

  get "/media-temple-cdn" do
    haml :index, :layout => :mediatemple
  end

  get "/microsoft-cdn" do
    haml :index, :layout => :microsoft
  end

  get "/themes/?" do
    haml :themes_list, :layout => :unspecified
  end

  get "/fallback/?" do
    haml :themes_list, :layout => :unspecified
  end

  
  get "/themes/" do
    haml :theme, :layout => :layout_theme
  end
  
  get "/themes/:theme" do |theme|
    @theme = theme
    haml :theme, :layout => :layout_theme
  end

  
  get "/fallback/" do
    haml :theme, :layout => :layout_fallback
  end

  get "/fallback/:theme" do |theme|
    @theme = theme
    haml :theme, :layout => :layout_fallback
  end
end

lib/rack/jquery_ui/themes/version.rb
module Rack
  class JQueryUI

    # This is here in case another rack-jquery_ui library has already been loaded.
    jquery_ui_version = "1.10.3"
    if defined? JQUERY_UI_VERSION
      warn "JQUERY_UI_VERSION was already defined."
      unless JQUERY_UI_VERSION == jquery_ui_version
        warn "The JQUERY_UI_VERSION defined is #{JQUERY_UI_VERSION} but the version this library wants to use is #{jquery_ui_version}. You have been warned!"
      end
    else
      # version of jQuery UI script
      JQUERY_UI_VERSION = jquery_ui_version

      # This is the release date of the jQuery file, it makes an easy "Last-Modified" date for setting the headers around caching.
      # @todo remember to change Last-Modified with each release!
      JQUERY_UI_VERSION_DATE = "Fri, 03 May 2013 00:00:00 GMT"
    end

    class Themes
      # version of this library
      VERSION = "3.0.1"
    end
  end
end


lib/rack/jquery_ui/themes.rb
require "rack/jquery_ui/themes/version"
require "rack/jquery/helpers"

# @see http://rack.github.io/
module Rack

  # @see https://github.com/yb66/rack-jquery_ui
  class JQueryUI
  
    # jQuery-UI themes' CDN script tags and fallback in one neat package.
    class Themes
      include JQuery::Helpers

      # The standard CSS file.
      JQUERY_UI_THEME_FILE = "jquery-ui.min.css"


      # Namespaced for convenience, to help with checking
      # which CDN supports what.
      module CDN
  
        # URL for the Media Temple CDN
        MEDIA_TEMPLE = "http://code.jquery.com/ui/#{JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/jquery-ui.min.css"
    
        # URL for the Google CDN
        GOOGLE = "//ajax.googleapis.com/ajax/libs/jqueryui/#{JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/jquery-ui.min.css"
    
        # URL for the Microsoft CDN
        MICROSOFT = "//ajax.microsoft.com/ajax/jquery.ui/#{JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/jquery-ui.min.css"

      end


      # This javascript checks if the jQuery-UI object has loaded by issuing a head request to the CDN. If it doesn't get a successful status, that most likely means the CDN is unreachable, so it uses the local jQuery-UI theme.
      FALLBACK = <<STR
<meta id='rack-jquery-ui-themes-fallback-beacon' />

<script type="text/javascript">
  var meta = $("#rack-jquery-ui-themes-fallback-beacon");
  meta.addClass("ui-icon");
  if ( meta.css('width') != "16px" ) {
    $('<link rel="stylesheet" type="text/css" href="/js/jquery-ui/#{JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/#{JQUERY_UI_THEME_FILE}" />').appendTo('head');
  }
  meta.remove();
</script>
STR

  
      # List of the standard themes provided by jQuery UI.
      STANDARD_THEMES = %w{black-tie blitzer cupertino dark-hive dot-luv eggplant excite-bike flick hot-sneaks humanity le-frog mint-choc overcast pepper-grinder redmond smoothness south-street start sunny swanky-purse trontastic ui-darkness ui-lightness vader}


      # @param [Hash] env The rack env hash.
      # @param [Hash] opts
      # @option opts [#to_sym] :organisation Choose which CDN to use, either :media_temple, :microsoft, or :media_temple.
      # @option opts [#to_s] :theme Theme to use. Won't set any routes or permanent settings, see note.
      # @option opts [TrueClass, Symbol] :fallback `true` if you want a fallback script, `false` if you don't, and `:only` if you don't want the CDN link but you do want just the fallback script. `true` is the default.
      # @return [String] The HTML script tags to get the CDN, with a JQuery function that will be called if the CDN fails and sets the fallback path.
      # @example
      #   # The easiest, use the defaults:
      #   Rack::JQueryUI::Themes.cdn env
      #
      #   # Choose a CDN to favour:
      #   Rack::JQueryUI::Themes.cdn env, :organisation => :media_temple
      #
      #   # Choose a theme to use:
      #   Rack::JQueryUI::Themes.cdn env, :organisation => :media_temple, theme: "dot-luv"
      #
      # @note
      #   The :theme option can override those given when
      #   setting up the middleware (see {#initialize}), but if the theme given here
      #   is not in the themes already set up, then there'll be
      #   no URL paths for it, so use this option to pick which
      #   theme is favoured *from the already given set*.
      def self.cdn( env, opts={} )
        organisation = opts[:organisation] || :media_temple
        themes = env["rack.jquery_ui-themes"]
        theme, themes = sort_out_options opts.merge :themes => themes
        fallback = opts[:fallback] || true

        # Get the CDN URL for the given organisation.
        url = case organisation.to_sym
          when :google then CDN::GOOGLE
          when :microsoft then CDN::MICROSOFT
          when :media_temple then CDN::MEDIA_TEMPLE
          else CDN::MEDIA_TEMPLE
        end

        script_name = env.fetch("SCRIPT_NAME","/")

        fallback_url = ::File.join script_name, "/js/jquery-ui/#{JQueryUI::JQUERY_UI_VERSION}/themes/:THEME/#{JQUERY_UI_THEME_FILE}"

        fallback_script = FALLBACK.gsub(/\:CDNURL/, url)
                                  .sub /\:FALLBACK_URL/, fallback_url

        if fallback == :only
          fallback_script.gsub(/\:THEME/, theme)
        else
          script = "<link rel='stylesheet' href='#{url}' type='text/css' />"
          if fallback == false
            script.gsub(/\:THEME/, theme)
          else
            [script,fallback_script].map{|x|
              x.gsub(/\:THEME/, theme)
            }.join("\n")
          end
        end
      end
  
  
      # Default options hash for the middleware.
      DEFAULT_APP_OPTIONS = {
        :http_path => "/js/jquery-ui/#{JQueryUI::JQUERY_UI_VERSION}/themes/"
      }


      # @private
      # Used internally to find the vendored files.
      Path_to_vendor = ::File.join(
        "../../../",
        "vendor/assets/javascripts/jquery-ui"
      )


      def self.sort_out_options( options={} )
        theme = options[:theme]
        themes = options[:themes]

        if theme
          t = Array(theme).map(&:to_s).uniq # put in an array
          themes = themes ?
            (t + themes).uniq :
            t
        else
          themes = ["smoothness"] if themes.nil? || themes.empty?
          theme = themes.first
        end
        
        unless (ts = themes - STANDARD_THEMES).empty?
          fail ArgumentError, %Q!The themes given (#{ts.join(" ")}) are unknown for this version of the rack-jquery_ui-themes library. Choose from #{STANDARD_THEMES}!
        end

        [theme,themes]
      end


      # @param [#call] app
      # @param [Hash] options
      # @option options [#to_s] theme The theme to use. The default is "smoothness".
      # @option options [Array<#to_s>] themes List of themes to allow. If the option for theme is set then this will be a backup set. If the option for theme is not set then the first theme listed will become the main theme.
      # @option options [String] :http_path If you wish the jQuery CSS fallback route to be "/js/jquery-ui/1.10.1/smoothness/jquery-ui.min.css" (or whichever version this is at) then do nothing, that's the default. If you want the path to be "/assets/javascripts/jquery-ui/1.10.1/smoothness…" then pass in `:http_path => "/assets/javascripts/#{Rack::JQueryUI::JQUERY_UI_VERSION}".
      # @note ***Don't leave out the version number when using the :http_path option!***. The scripts provided by jQuery don't contain the version in the filename like the jQuery scripts do, which means that organising them and sending the right headers back is bound to go wrong unless you put the version number somewhere in the route. You have been warned!
      # @example
      #   # The default:
      #   use Rack::JQueryUI::Themes
      #
      #   # With a default theme other than smoothness:
      #   use Rack::JQueryUI::Themes, :theme => "dot-luv"
      #
      #   # With a several themes, the main one being dot-luv:
      #   use Rack::JQueryUI::Themes, :themes => %w{dot-luv blitzer eggplant trontastic}
      #
      #   # Another way to do the same thing
      #   use Rack::JQueryUI::Themes, :theme => "dot-luv", :themes => %w{blitzer eggplant trontastic}
      def initialize( app, options={} )
        @app, @options  = app, DEFAULT_APP_OPTIONS.merge(options)
        theme, @themes = self.class.sort_out_options @options

        @http_base_path = @options[:http_path]

        # set the path to all the assets
        @paths = @themes.each_with_object({}) do |theme,paths|
          version_and_theme = ::File.join( 
            JQueryUI::JQUERY_UI_VERSION, "themes", theme
          )

          theme_url = ::File.join @http_base_path, theme
          theme_dir  = ::File.expand_path( ::File.join( Path_to_vendor, version_and_theme ), ::File.dirname(__FILE__) )

          # Store the HTTP path to the CSS as the key
          # with the file path to the CSS file as the value
          paths.store(
            ::File.join( theme_url, JQUERY_UI_THEME_FILE ),
            {
              :file => ::File.join( theme_dir, JQUERY_UI_THEME_FILE),
              :theme => theme
            }
          )

          images_dir = ::File.join theme_dir, "images"

          Dir.new(images_dir)
              .entries
              .delete_if{|f| f =~ /^\./ }
              .select{|f| 
                Rack::Mime.mime_type(
                  ::File.extname(f), "text/css" # fallback is a false
                ).start_with? "image"           # here
              }.each do |img|
                paths.store(
                  ::File.join( theme_url, "images", img ), # url
                  {
                    :file   =>  ::File.join( images_dir, img ),
                    :theme  =>  theme
                  }
                )
              end
        end
      end


      # @param [Hash] env Rack request environment hash.
      def call( env )
        dup._call env
      end
  
  
      # For thread safety
      # @param (see #call)
      def _call( env )
        env.merge! "rack.jquery_ui-themes" => @themes
        request = Rack::Request.new(env.dup)

        # path for CSS
        # TODO benchmark this condition
        # start_with? vs has_key vs regex
        if @paths.has_key? request.path_info
          response = Rack::Response.new
          if request.path_info.end_with? JQUERY_UI_THEME_FILE # CSS.
            # For caching:
            response.headers.merge! caching_headers("#{JQUERY_UI_THEME_FILE}-#{@paths[request.path_info][:theme]}-#{JQueryUI::JQUERY_UI_VERSION}", JQueryUI::JQUERY_UI_VERSION_DATE)
            # There's no need to test if the IF_MODIFIED_SINCE against the release date because the header will only be passed if the file was previously accessed by the requester, and the file is never updated. If it is updated then it is accessed by a different path.
            requested_file = open @paths[request.path_info][:file], "r"
          elsif request.path_info =~ /images/
            # serve images
            requested_file = open @paths[request.path_info][:file], "rb"
          else # bad route
            response.status = 404
            return response.finish # finish early, 404
          end

          # set status and read the requested file
          if request.env['HTTP_IF_MODIFIED_SINCE']
            response.status = 304
          else
            response.status = 200
            mime = Mime.mime_type(::File.extname(requested_file.path), 'text/css')
            response.headers.merge!( "Content-Type" => mime ) if mime
            response.write ::File.read( requested_file )
          end
          response.finish
        else
          @app.call(env)
        end
      end # call
  
  
    end # Themes
  end
end
